python/xorbits/_mars/services/task/analyzer/fusion.py

```diff
@@ -109,8 +113,17 @@ def _color_init_nodes(self) -> Dict[OperandType, int]:
     def color(self) -> Dict[ChunkType, int]:
         chunk_to_colors = dict()
 
-        # step 1: Coloring the initial nodes according to the bands that assigned by assigner
-        op_to_colors = self._color_init_nodes()
+        # step 1: Coloring the initial nodes according to the slots that assigned by assigner
+        # if the number of initial color is less than the number of slots,
+        # decrease initial_same_color_num until the number of initial color exceeds slots.
+        slots_num = sum((r.num_cpus + r.num_gpus) for r in self.band_resource.values())
+        while self.initial_same_color_num > 0:
+            self.reset_color()
+            op_to_colors = self._color_init_nodes()
+            if len(set(op_to_colors.values())) >= slots_num:
+                break
+            else:
+                self.initial_same_color_num -= 1
```

Comment:
Nice, this logic handle most cases. Can we simply make the data source init nodes embarassing parallel?
Pros: all the data sources are parallel executed, minimize the impact of slow workers / slots.
Cons:  the fusion is bad which makes the subtask graph large.

Reply:
This may require some benchmarking, but currently I prefer to keep the possibility of fusion.